add notes for n25600 (#551)
bitcoin-core-review-club/website: _posts/2022-08-31-#25600.md

---
layout: pr
date: 2022-08-31
title: "Advertise `NODE_REPLACE_BY_FEE` and connect to 1 outbound full-rbf peer if `mempoolfullrbf` sets"
pr: 25600
authors: [ariard]
components: ["p2p"]
host: mzumsande
status: upcoming
commit:
---

> ## Notes

Author: done.

> - [BIP 125](https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki) (Replace-by-fee)

Reviewer: ```suggestion
- [Bitcoin Core's RBF policy](https://github.com/bitcoin/bitcoin/blob/23.x/doc/policy/mempool-replacements.md#mempool-replacements)
```

Author: done.

enables unconfirmed transactions to be replaced by conflicting transactions ("double spends") under certain
conditions (see the BIP for details). If these conditions are met, nodes will replace the transaction
in their mempool and relay the replacing transaction to their peers.

> - One important feature of BIP 125 is that it is opt-in: Only original transactions that signal

Reviewer: ```suggestion
- The default policy requires opt-in [BIP 125 signaling](https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki): Only original transactions that signal
```

Author: done

replaceability (using the `nSequence` field on one of its inputs) are subject to replacement.
If the original transaction didn't signal replaceability, the replacement transaction
will be neither accepted nor relayed.

- For L2 protocols, the necessity to opt-in to RBF poses DoS problems, see
[this mailing list post](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-June/020557.html)
for a summary of research.
Therefore, it has been suggested to make full-RBF possible (removing the necessity to opt in).

- As a first step, [PR 25353](https://github.com/bitcoin/bitcoin/pull/25353) added an optional
`-mempoolfullrbf` parameter to Bitcoin Core, which is disabled by default.
If the node operator enables this, the node will ignore the signaling (but will still adhere to other RBF rules).

- This PR goes one step further by implementing a preferential peering solution for peers that

> have `-mempoolfullrbf` enabled. These peers would now signal full-RBF and make additional

Author: changed, thanks.

> automatic outbound connections to peers that have also signaled full-RBF.

Reviewer: Might be worth mentioning there is no increase in overall connections because bandwidth is limited, hence "preferential" peering, not just additional peers "for free."

Author: Are you sure? Looking at the code, I think it's the latter, and it's meant to be N_RBF additional outbound connections (which currently doesn't to work as intended, I'll leave a comment in the PR), so it would be 8+2+N_RBF outbound connections.
Of course, this means less inbound slots, but for a non-listening node or a node that is not at the maximum of its inbound connections, it should mean an increase in connections and traffic.
What do you mean with "bandwidth is limited"?

Reviewer: Nevermind sorry, resolving. I just meant to say that increasing overall peers would increase bandwidth usage. I haven't looked closely but for some reason I assumed the PR wasn't adding extra outbound slots, oops.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

> 1. What network-level conditions must be fulfilled for full-RBF to "work"? Is it necessary

Reviewer: ```suggestion
1. What network-level conditions must be fulfilled for full-RBF to "work," i.e., facilitate the propagation of transactions that replace others not signaling BIP 125? Is it necessary
```

Author: done
